### PR TITLE
fix(ApplicationActivation): Remove registration client roles

### PR DESCRIPTION
**Pull request by Phil91**
Removes the companyUser assigned roles for a specific clientId while activating the company

Refs: [TEST-522](https://jira.catena-x.net/browse/TEST-522)

### CODE REVIEW

File: src/checklist/Checklist.Worker/Program.cs [hunk 1]
```diff
@@ -46,12 +48,20 @@
...
          .AddChecklist(hostContext.Configuration.GetSection("Checklist"))
          .AddChecklistCreation()
-         .AddApplicationActivation(hostContext.Configuration)).Build();
+         .AddApplicationActivation(hostContext.Configuration);
+      
+      var urlsToTrust = hostContext.Configuration.GetSection("Keycloak").Get<KeycloakSettingsMap>().Values
+         .Where(config => config.ConnectionString.StartsWith("https://"))
+         .Select(config => config.ConnectionString)
+         .Distinct();
+      FlurlUntrustedCertExceptionHandler.ConfigureExceptions(urlsToTrust);
```
Comment (ntruchsess) [inline src/checklist/Checklist.Worker/Program.cs:63]: for a production-system it's is a security-issue to ignore untrustworthy certificates, so this must be configurable to disable for non-dev environments.

File: src/registration/ApplicationActivation.Library/ApplicationActivationService.cs [hunk 2]
```diff
@@ -172,6 +174,22 @@ await _provisioningManager
...
+    private async Task RemoveRegistrationRoles(Guid applicationId, IUserRolesRepository userRolesRepository, IApplicationRepository applicationRepository)
+    {
+        var iamClientIds = _settings.ClientToRemoveRolesOnActivation;
+        var invitedUsersData = applicationRepository
+            .GetUserDataForRoleDeletionByIamClientIds(applicationId, iamClientIds);
+        await foreach (var userData in invitedUsersData.ConfigureAwait(false))
+        {
+            if (!userData.RolesToDelete.Any()) continue;
+
+            var roleNamesToDelete = iamClientIds.ToDictionary(clientId => clientId, _ => userData.RolesToDelete.Select(x => x.UserRoleText));
```
Comment (ntruchsess) [inline src/registration/ApplicationActivation.Library/ApplicationActivationService.cs:186]: What do we need a dictionary for if keys are assigned to the same value? Technically the provisionmanager expects the dictionary but assigning the same roles to all clients ain't work for multiple clientIds.
To get the correct mapping the query needs to include the ClientClientId a CompanyUserRole is associated with in the result.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/checklist/Checklist.Worker/Program.cs (hunk 1 above)
Comment (Phil91) [reply src/checklist/Checklist.Worker/Program.cs]: check for dev env added

File: src/registration/ApplicationActivation.Library/ApplicationActivationService.cs (hunk 2 above)
Comment (Phil91) [reply src/registration/ApplicationActivation.Library/ApplicationActivationService.cs]: ClientClientId added to the userRoles

**Comment by Phil91** [comment]
I`ve fixed the ApplicationActivation UnitTests besides of adding a new one for the deletion.
The merge now will merge the PR against #138 not main

File: src/portalbackend/PortalBackend.DBAccess/Repositories/ApplicationRepository.cs [hunk 3]
```diff
@@ -353,12 +353,22 @@ public void AttachAndModifyCompanyApplication(Guid companyApplicationId, Action<
...
+             .Where(companyUser => companyUser!.CompanyUserStatusId == CompanyUserStatusId.ACTIVE)
+             .Select(companyUser => new UserRoleDeletionData(
+                 companyUser!.Id,
+                 companyUser.IamUser!.UserEntityId,
+                 companyUser.CompanyUserAssignedRoles
+                     .Where(x => x.UserRole!.Offer!.AppInstances.Any(ai => iamClientIds.Contains(ai.IamClient!.ClientClientId)))
+                     .Select(companyUserAssignedRole => new ValueTuple<string, Guid, IEnumerable<string>>(
+                         companyUserAssignedRole.UserRole!.UserRoleText,
+                         companyUserAssignedRole.UserRoleId,
+                         companyUserAssignedRole.UserRole!.Offer!.AppInstances.Select(x => x.IamClient!.ClientClientId)
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/ApplicationRepository.cs:371]: as this is highly redundant in the db-resultset it's probably more efficient to run separate queries. One that retrieves the roles with related data (id, userRoleText, clientClientIds) and one for users+assigned roleids in separate queries.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/administration/Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 4]
```diff
@@ -317,7 +317,7 @@ public async Task ProcessClearinghouseResponseAsync(string bpn, ClearinghouseRes
     }
 
     /// <inheritdoc />
-    public Task SetRegistrationVerification(Guid applicationId, bool approve, string? comment)
+    public Task SetRegistrationVerification(Guid applicationId, bool approve, string? comment = null)
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/administration/Administration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: changed.

File: src/portalbackend/PortalBackend.DBAccess/Repositories/ApplicationRepository.cs (hunk 3 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/ApplicationRepository.cs]: changed the query to two queries.

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**